ktor-client/ktor-client-cio/jvmAndPosix/src/io/ktor/client/engine/cio/ConnectionFactory.kt

```diff
@@ -39,8 +60,63 @@ internal class ConnectionFactory(
         }
     }
 
-    fun release(address: InetSocketAddress) {
-        addressLimit[address]!!.release()
-        limit.release()
+    fun release(socket: Socket) {
+        LOG.trace { "Releasing connection for address: ${socket.remoteAddress}" }
+
+        if (socket.isClosed) {
+            LOG.warn("Attempted to release a closed connection for address: ${socket.remoteAddress}")
+            return
+        }
+
+        val pooledConnection = PooledConnection(socket, GMTDate().timestamp)
+        synchronized(this) {
+            val pool = connectionPool.getOrPut(socket.remoteAddress) { mutableListOf() }
+            pool.add(pooledConnection)
+
+            // If pool size exceeds maxPoolSize, remove and close excess connections
+            while (pool.size > maxPoolSize) {
+                LOG.trace { "Pool size exceeded maxPoolSize, removing oldest connection" }
+                pool.removeAt(0).close()
+            }
+
+            limit.release() // Release the semaphore to allow new connections
+            addressLimit[socket.remoteAddress]?.release() // Release the address semaphore
+        }
+        LOG.trace { "Connection released for address: ${socket.remoteAddress}" }
+    }
+
+    private fun getPooledConnection(address: InetSocketAddress): PooledConnection? {
+        LOG.trace { "Checking for pooled connection for address: $address" }
+        val connections = connectionPool[address] ?: return null
+        val currentTime = GMTDate().timestamp
+
+        synchronized(this) {
+            connections.removeAll {
+                if (currentTime - it.lastUsed > keepAliveTime) {
+                    LOG.trace { "Removing expired connection for address: $address" }
+                    it.close()
+                    return@removeAll true
+                }
+                false
```

Comment:
Should we also remove closed connections here? Is it possible to have a closed connection in the pool?